### ISSUE TITLE
Can't update non published entries

### ISSUE THREAD

**Issue by pygillier** [bolt/bolt#3466]
When I try to update an entry (default content-type) which is in draft state, the updates can't be saved.

After checking and debugging, it seems to be related to endpoint.

In file src/Storage.php, line 1245: check is done on status & endpoint : status is unset and $app['end'] is set to async. So, the condition will validate and set status in query to "status = published".

So, no object will be returned and update will fail.

Tested on 2.1.8 and it seems to be linked to this commit : cfbb0ec9d51563bf3a117acf98d395be61b71b54

**Comment by bobdenotter** [comment]
@pygillier Looking at it, i think you're right. Expect a fix soon.

**Comment by bobdenotter** [comment]
Fix can be found here: #3467

**Comment by pygillier** [comment]
Hi, 

is there a ETA on releasing this patch ? ATM, I can't use my instance because of this bug.

**Comment by fabschurt** [comment]
This problem is still present in `v2.2.4`. I doubt that 0037c11 or 85fe99c are related to this problem at all (even though I concede that I'm having a hard time figuring out the big picture here). All I know is that if I revert cfbb0ec, it solves the problem.

I'm kind of puzzled by cfbb0ec. Its title reads **Make sure we set the status correctly for 'async' requests.**, so I'd expect the commit to fix a problem related to the `async` end, but AFAIK it introduces a bug instead. Moreover, is there actually a point in forcing the `published` status criteria for other ends than `frontend`?

**Comment by bobdenotter** [comment]
Hi,

I assumed this was fixed. I installed a fresh new Bolt 2.2.4 to try it, and I can't reproduce this behaviour. This is what I did: 

![screenshot_2015-07-21_19_51_25](https://cloud.githubusercontent.com/assets/1833361/8808351/602ce4f4-2fe2-11e5-9dfa-123de51a5c4a.png)

![screenshot_2015-07-21_19_51_43](https://cloud.githubusercontent.com/assets/1833361/8808354/63f5a364-2fe2-11e5-867e-15f84ad079e0.png)

![screenshot_2015-07-21_19_51_55](https://cloud.githubusercontent.com/assets/1833361/8808355/66f1e1fe-2fe2-11e5-9b91-d25dafc8becf.png)

Are you sure it's not working for you? If so, could you give me more details on how to reproduce it exactly? :-)

**Comment by fabschurt** [comment]
Hey @bobdenotter, thanks for taking the time to look into this. I took the time to start from a fresh `v2.2.4` install too, to make sure that none of my various overrides was the cause of the problem. I just copy-pasted my `contenttypes.yml` and `permissions.yml` files from my current project:

**`contenttypes.yml`:**

```yaml
__nodes:
  default_fields: &default_fields
    title:
      type: text
      group: content
      required: true
    slug:
      type: slug
      group: content
      uses: title
    content:
      type: html
      group: content
      height: 500px

pages:
  name: Pages
  singular_name: Page
  slug: pages
  singular_slug: page
  icon_one: fa:file-text-o
  icon_many: fa:file-text-o
  fields:
    <<: *default_fields
    photos:
      type: imagelist
      group: attachments
      upload: photos
    files:
      type: filelist
      group: attachments
      upload: fichiers

sessions:
  name: Sessions de séjour
  singular_name: Session de séjour
  slug: sejours
  singular_slug: sejour
  icon_one: fa:calendar
  icon_many: fa:calendar
  relations:
    articles:
      label: Assign blog posts
      order: -datepublish
      multiple: true
  fields:
    title:
      type: text
      group: content
      required: true
    slug:
      type: slug
      group: content
      uses: title
    password:
      type: text
      group: content
      required: true
    introduction:
      type: html
      group: content
      height: 300px

posts:
  name: Articles de blog
  singular_name: Article de blog
  slug: articles
  singular_slug: article
  icon_one: fa:pencil
  icon_many: fa:pencil
  searchable: false
  relations:
    sejours:
      label: Assign to session
      multiple: false
  fields:
    <<: *default_fields
    photos:
      type: imagelist
      group: attachments
      upload: photos-sejours

elements:
  name: Éléments
  singular_name: Élément
  slug: elements
  singular_slug: element
  default_status: published
  icon_one: fa:cog
  icon_many: fa:cog
  searchable: false
  viewless: true
  fields:
    <<: *default_fields
```

**`permissions.yml`:**

```yaml
roles:
  admin:
    description: 'This user can pretty much do anything but is not root for security reasons.'
    label: Administrator
  manager:
    description: 'This user holds all rights over content editing and users, but cannot touch the website configuration.'
    label: Manager
  editor:
    description: 'This user can create and edit their own posts.'
    label: Editor

roles-hierarchy:
  manipulate:
    admin: [admin]
    manager: [admin]
    editor: [admin, manager]

global:
  about: [admin]
  clearcache: [admin]
  contentaction: [admin, manager, editor]
  dashboard: [everyone]
  dbcheck: [admin]
  dbupdate: [admin]
  dbupdate_result: [admin]
  extensions: [admin]
  extensions:config: []
  fileedit: []
  files:config: []
  files:theme: []
  files:uploads: [admin, manager, editor]
  files: [admin, manager, editor]
  prefill: []
  profile: [everyone]
  settings: [admin, manager]
  translation: []
  useraction: [admin, manager]
  useredit: [admin, manager]
  users: [admin, manager]
  roles: [admin]
  maintenance-mode: [admin]
  omnisearch: [admin, manager]
  changelog: [admin]
  systemlog: [admin]
  login: [anonymous]
  postLogin: [anonymous]
  logout: [everyone]

contenttype-all:
  edit: [admin]
  create: [admin]
  publish: [admin]
  depublish: [admin]
  delete: [admin]
  change-ownership: [admin]
  view: [admin]

contenttype-default:
  edit: [manager, owner]
  create: []
  publish: [manager, owner]
  depublish: [manager, owner]
  delete: [owner]
  change-ownership: []
  view: [manager]

contenttypes:
  sejours:
    create: [manager, editor]
    delete: [manager]
    change-ownership: [manager]
    view: [manager, editor]
  articles:
    create: [manager, editor]
    delete: [manager, owner]
    view: [manager, editor]
```

Here are the exact steps to reproduce the bug:

1. Create one user for each available roles (`admin`, `manager`, `editor`).
2. As the `admin` or `manager`, create a `sejours` entity, and set the author to be the `editor`.
3. The `editor` is now the owner of the `sejours` entity, and can edit its content.
4. As the `editor`, apply some changes to the created `sejours` entity (`http://bolt.dev/bolt/editcontent/sejours/1`) and save.
5. After the changes seem to be saved, refresh the page or go back to the overview page (`http://bolt.dev/bolt/overview/sejours`), and see that the changes have not been saved.

Editing the `sejours` entity as `root`, `admin` or `manager` works just fine. The problem only occurs when editing as `editor`. I activated `debug_permission_audit_mode` in `config.yml` to get some input; here are the (nicely formatted) log rows produced when trying to save some changes for a `sejours` entity as an `editor`:

```
2015-07-21 23:33:41 - Checking permission query 'login' for user 'editor' with contenttype '' and contentid ''
2015-07-21 23:33:41 - "Granting "login" based on role anonymous"
2015-07-21 23:33:41 - Checking permission query 'editcontent' for user 'editor' with contenttype '' and contentid ''
2015-07-21 23:33:41 - Granting 'editcontent' (hard-coded override)
2015-07-21 23:33:41 - Checking permission query 'contenttype:sejours:edit:1' for user 'editor' with contenttype '' and contentid ''
2015-07-21 23:33:41 - "Denying "edit" for sejours; available roles: editor, everyone, anonymous"
2015-07-21 23:33:41 - Checking permission query 'dashboard' for user 'editor' with contenttype '' and contentid ''
2015-07-21 23:33:41 - "Granting "dashboard" based on role everyone"
```

Note: as an `editor`, changing the status of a `sejours` entity from the overview listing (with the `Actions` dropdown) remains possible.

I'm confused. I've tried to investigate `Bolt\Controllers\Backend`, `Bolt\Users`, `Bolt\Permissions` and `Bolt\PermissionParser`, but I'm having some trouble figuring it all out.

Would you have any idea?

**Comment by fabschurt** [comment]
Hi @bobdenotter, sorry to bother you about this issue, but have you (or anyone else) been able to reproduce the bug using the context I provided above? I've been working on other things so I wasn't able to research this further, but I'm gonna work on it again sometime in the next few days, so any insight would be of great help!

**Comment by fabschurt** [comment]
So, I've updated Bolt to `2.2.8`, and now it works like a charm. I don't know how, I don't know why, but it works, so that's pretty cool :sweat_smile:

**Comment by fabschurt** [comment]
All right, so, I've found the origin of my problem: apparently it's a duplicate of #3909, sorry for all the trouble.

**Comment by fabschurt** [comment]
@bobdenotter, @GawainLynch, well in the end, this has nothing to do with #3909, the problem still occurs, even after pulling the latest changes from `release/2.2`. Here's a detailed description of the context, in order to reproduce the bug:

* I use `dev-release/2.2` in `composer.json`, up to date with the upstream.

* I try to save changes to a record (slug: `articles`) with `held` status as a user who identifies as the `owner` of the record (role: `editor`), and has no `edit` capabilities concerning its `contenttype` otherwise. My `contenttypes.yml` looks like this:

```yaml
contenttype-default:
  edit: [manager, owner]
  create: []
  publish: [manager, owner]
  depublish: [manager, owner]
  delete: [owner]
  change-ownership: []
  view: [manager]
contenttypes:
  articles:
    create: [manager, editor]
    delete: [manager, owner]
    view: [manager, editor]
```

* **Note:** if I try to save changes as a user with real `edit` capabilities over the `contenttype` (a `manager` for example), it works. This won't help us here, but it's just to emphasize the fact that the bug can only be reproduced when the user is the `owner` of the record and has no `edit` capabilities over its `contenttype` otherwise.

* So, on the edit page (`/admin/editcontent/articles/{id}`), when I save the changes, there's no error message displaying, and everything seems fine, but when I reload the page, I can see that none of the changes have been saved.

The problem comes from [here](https://github.com/bolt/bolt/blob/release/2.2/src/Storage.php#L1310). This part of the code forces the retrieved post's status to `published` for all app ends but `backend`. The problem is, when we save changes to a record on its edit page in the admin area, the app end is actually `async`, and not `backend`. So in that case, `Bolt\Storage::getContent()` returns `false` as the record with a `held` status can't be found, and the user can't be given the `owner` role (as normally expected) [here](https://github.com/bolt/bolt/blob/release/2.2/src/Permissions.php#L673).

This problem was introduced with ae1f672. By reverting this commit, it removes the problem, but there must be a legitimate reason to apply this forced status to the `async` end as well I guess?

**Comment by GawainLynch** [comment]
>  By reverting this commit, it removes the problem, but there must be a legitimate reason to apply this forced status to the async end as well I guess?

There was: #3463.

**Comment by fabschurt** [comment]
Well, yeah, but there's not much of a discussion about why `published` is the theoretically _correct_ status for `async` requests. I'm confused. The thing is that it breaks the editing functionality in the context I described above.

But anyway, this can actually be resolved by changing [these lines](https://github.com/bolt/bolt/blob/release/2.2/src/Permissions.php#L668-L670) in `Bolt\Permissions::isAllowedSingle()` to:

```php
if (is_string($content) || ($contenttype && $contentId)) {
    $content = $this->app['storage']->getContent("$contenttype/$contentId", array('hydrate' => false, 'status' => '!undefined'));
}
```

That would be acceptable?

**Comment by GawainLynch** [comment]
Sorry if my response sounded like a deflection… It *should have* come across as an "in passing" comment while waiting on @bobdenotter to chime in as I honestly can't remember all the details and his memory beats mine every time! :grin:

**Comment by fabschurt** [comment]
Oh no, it didn't sound like a deflection at all, and I'm sure there's a 100% legitimate reason why this has been done this way, that's why I think the solution I proposed would be the best way to go, as I've seen this "trick" being used in various places of `Bolt\Controllers\Backend` as well, so I guess it's a pretty "standard" procedure. Let's wait for @bobdenotter and his uber-memory then :+1:. I'll prepare a PR in the meantime, just in case.

**Comment by GawainLynch** [comment]
Cool… One thing that is hampering me is that I had a lot to do with replacing the request handling for the upcoming v2.3 `master` branch and, well, a lot has changed so not the best person to comment… and I am trying to not unleash the killer :koala:
